SU2_CFD/include/variables/CNEMOEulerVariable.hpp

```diff
@@ -333,19 +350,35 @@ class CNEMOEulerVariable : public CVariable {
   bool SetPrimVar(unsigned long iPoint, CFluidModel *FluidModel) override;
 
  /*!
-  * \brief Set all the conserved variables.
+  * \brief Set all the primitive and secondary variables from the conserved vector.
   */
   bool Cons2PrimVar(su2double *U, su2double *V, su2double *dPdU,
                     su2double *dTdU, su2double *dTvedU, su2double *val_eves,
                     su2double *val_Cvves);
 
+ /*!
+  * \brief Set all the conserved variables from the primitive vector..
+  */
+  void Prim2ConsVar(su2double *U, su2double *V);
+
  /*!
   * \brief Check for unphysical points.
   * \return Boolean value of physical point 
   */
-  bool CheckNonPhys(su2double *U, su2double *V, su2double *dPdU,
-                    su2double *dTdU, su2double *dTvedU, su2double *val_eves,
-                    su2double *val_Cvves);
+  bool CheckNonPhys(su2double *V);
+
+  /*!
+   * \brief Recompute the extrapolated quantities, after MUSCL reconstruction,
+   *        in a more thermodynamically consistent way.
+   * \param[in] V - primitve variables.
+   * \param[out] d*dU - reconstructed secondaryvariables.
+   * \param[out] val_eves - reconstructed eve per species.
+   * \param[out] val_cvves - reconstructed cvve per species.
+   * \param[out] Gamma - reconstructed gamma.
+   */
+  su2double ComputeConsistentExtrapolation(su2double *V, su2double* dPdU,
+                                           su2double* dTdU, su2double* dTvedU,
+                                           su2double* val_eves,su2double* val_cvves);

```

Comment:
When something is an input passed by reference or pointer it should be const (likewise for the method itself if it does not change the object).
This function caught my eye because it has both a return and output parameters, and because it does not have iPoint as input, which maybe suggests this can be static? Or that V is not needed because it is already there?

Reply:
Thanks for taking a look! I appreciate the feedback and advice :)

I wasn't sure how to best handle this function.  I could pass gamma by reference as well.   That being said, I am admittedly lacking in the proper use of "static" functions.   The V vector is used to compute the dPdU, etc after its been recomputed in the primitive limiter schemes.

Please correct me if I am wrong here, V can be passed as const,  but since dPdU, eves, etc are computed in the function, they cant be const?

Reply:
Looking at it more closely, it does not belong in the variable class.
CVariable should be a container, if it starts being used to compute physical quantities, later down the line there will be much philosophical discussion about where to implement new things.
Granted we don't always follow this advice of "just a container" (perhaps I should fix that) but I think the general idea should be:
- CVariable: Just a container, the only info it should deal with is what variables to allocate (and how).
- CSolver: Orchestration, i.e. looping over points / edges doing very generic things.
- "Models": e.g. CNumerics, CFluidModel, and other physical models we should probably have but don't (as @clarkpede was suggesting as few days ago). These are parts of the solver that can be changed to make it do different things. These models do not have to be always classes, they can be free functions (like the ones used to compute gradients), or static methods of some class (static methods are called on the class rather than on the object, basically they are functions).

For this case in particular the thing to do would be to move these functions to the solver, I think there is some precedent for that in CEulerSolver with the consistent reconstruction stuff.

Reply:
That makes sense, Im onboard with that logic.